src/zimcheck.cpp

```diff
@@ -396,28 +396,35 @@ void test_articles(const zim::File& f, ErrorLogger& reporter, ProgressBar progre
             baseUrl.resize( pos==baseUrl.npos ? 0 : pos );
 
             auto links = getLinks(it->getData());
-            for(auto olink: links)
+            std::vector<std::string> filtered;
+            for (const auto &l : links)
             {
-                if (olink.front() == '#' || olink.front() == '?')
-                    continue;
-                if (isInternalUrl(olink)) {
-                    auto link = normalize_link(olink, baseUrl);
-                    char nm = link[0];
-                    std::string shortUrl(link.substr(2));
-                    auto a = f.getArticle(nm, shortUrl);
-                    if (!a.good())
+                if (l.front() == '#') continue;
+                if (isInternalUrl(l) == false) continue;
+
+                filtered.push_back(normalize_link(l, baseUrl));
+            }
+
+            // remove duplicates
+            sort(filtered.begin(), filtered.end());
+            filtered.erase(unique(filtered.begin(), filtered.end()), filtered.end());
+            for(const auto &olink: filtered)
+            {
+                char nm = olink[0];
+                std::string shortUrl(olink.substr(2));
+                auto a = f.getArticle(nm, shortUrl);
+                if (!a.good())
+                {
+                    int index = it->getIndex();
+                    if ((previousLink != olink) && (previousIndex != index) )
                     {
-                        int index = it->getIndex();
-                        if ((previousLink != link) && (previousIndex != index) )
-                        {
-                            std::ostringstream ss;
-                            ss << link << " (" << olink << ") was not found in article " << it->getLongUrl();
-                            reporter.addError(URL_INTERNAL, ss.str());
-                            previousLink = link;
-                            previousIndex = index;
-                        }
-                        reporter.setTestResult(URL_INTERNAL, false);
+                        std::ostringstream ss;
+                        ss << " (" << olink << ") was not found in article " << it->getLongUrl();
```

Comment:
We should display the link as it is written in the html. Else it may be difficult for the user to investigate and fix the issue.